### PR TITLE
Fix for compare method not rejecting promise on error

### LINKED ISSUE #468: Compare method not rejecting promise on error
Hello,

First of all, it's great that bcrypt now supports a Promise API.
However, it seems there are still a few "holes" in it.

One that I ran into, was related to these lines of code:

https://github.com/kelektiv/node.bcrypt.js/blob/master/bcrypt.js#L157
https://github.com/kelektiv/node.bcrypt.js/blob/master/bcrypt.js#L163

If there is no callback provided (e.g. Promise API is used), then both of these lines will trigger a `TypeError: cb is not a function`.

What needs to happen, is a similar check as a few lines below, and then reject with the error instead of passing it to the callback, if no callback is given. Something like this:

```js
const err = new Error('data and hash arguments required');
if (!cb) {
  return Promise.reject(err);
}
else {
  return process.nextTick(function() {
      cb(err);
  });
}
```

I would create a PR, but not sure what the deal is with wrapping the errors in `process.nextTick`.

This is with brypt 1.0.1.

### PR DESCRIPTION
**Pull request by Agathver**
This pull request aims to fix #468 - Compare method not rejecting promise on error

### CODE REVIEW

**Comment by defunctzombie** [comment]
Needs tests (didn't notice if those were added due to the diff noise)

**Comment by Agathver** [comment]
I removed the unnecessary changes.

File: bcrypt.js [hunk 1]
```diff
@@ -152,15 +152,15 @@ module.exports.compareSync = function compareSync(data, hash) {
 /// @param {String} hash expected hash
 /// @param {Function} cb callback(err, matched) - matched is true if hashed data matches hash
 module.exports.compare = function compare(data, hash, cb) {
-    if (data == null || hash == null) {
+    if (typeof data === 'function') {
         return process.nextTick(function() {
-            cb(new Error('data and hash arguments required'));
+            data(new Error('data and hash arguments required'));
```
Comment (defunctzombie) [inline bcrypt.js]: this doesn't really match the `if` condition. The user might provide a function for the data argument and then see this error when in fact they did provide an argument.

This also doesn't handle the case where the user passes null for the arguments and does not specify a callback. They will not get a rejection.

The real mistake of the API is to not throw then it is called with the wrong arguments which is what should have happened but given that is not the case. Fixing this to match previous behavior means identifying the error condition and then rejecting with promise or callback as needed.

Comment (Agathver) [inline bcrypt.js]: I tried to make it identical to `hash`, where it is assumed that if the user passes a function in place of parameters, the data and salt arguments are missing and the passed function is the callback which should be called with the error.

May be the error message could be rewritten as `'data and hash must be strings'` ?

If the user passes null for either the data or hash, and does not specify a callback, the Promise is rejected.

The previous (v0.8.7) behavior was to fail with `TypeError: cb is not a function`, But `hash` used to call the passed method and silently fail if no callback was specified.

Comment (defunctzombie) [inline bcrypt.js]: Understood.

Comment (adamreisnz) [inline bcrypt.js]: There is no check now in this function for `(data == null || hash == null)` when no callback is provided (also in the hash function). Is this not going to be an issue?

Comment (defunctzombie) [inline bcrypt.js]: When no callback is provided this function returns a promise (which is made simply from providing a callback and calling again) and then the null checks are present and the promise is rejected.

Comment (adamreisnz) [inline bcrypt.js]: You're right, I missed that it's simply calling itself again. Thanks!

**Pull request merged**